### PR TITLE
Prototype histogram advice API (i.e. Hints)

### PR DESCRIPTION
**Pull request by jack-berg**
A prototype extension to the metric API to allow instrumentation authors to optionally provide advice on histogram bucket boundaries. Advice only applies if there are no matching views and if the reader has no preference for the histogram bucket boundaries. 

Here's a code snippet that demonstrates usage and behavior:
```
    InMemoryMetricReader reader = InMemoryMetricReader.create();
    meterProvider = SdkMeterProvider.builder().registerMetricReader(reader).build();
    DoubleHistogram doubleHistogram =
        ((ExtendedDoubleHistogramBuilder) meterProvider
            .get("meter")
            .histogramBuilder("histogram"))
            .setAdvice(
                advice -> advice.setExplicitBucketBoundaries(Arrays.asList(10.0, 20.0, 30.0)))
            .build();

    doubleHistogram.record(5.0);
    doubleHistogram.record(15.0);
    doubleHistogram.record(25.0);
    doubleHistogram.record(35.0);

    // Bucket bounds from advice should be used
    assertThat(reader.collectAllMetrics())
        .satisfiesExactly(
            metric ->
                assertThat(metric)
                    .hasName("histogram")
                    .hasHistogramSatisfying(
                        histogram ->
                            histogram.hasPointsSatisfying(
                                point ->
                                    point
                                        .hasBucketCounts(1, 1, 1, 1)
                                        .hasBucketBoundaries(10.0, 20.0, 30.0))));
```

### CODE REVIEW

File: sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/AdviceTest.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import io.opentelemetry.sdk.metrics.export.DefaultAggregationSelector;
+import io.opentelemetry.sdk.testing.exporter.InMemoryMetricReader;
+import java.util.Arrays;
+import java.util.Collections;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+
+class AdviceTest {
+
+  private SdkMeterProvider meterProvider = SdkMeterProvider.builder().build();
```
Comment (mateuszrzeszutek) [inline sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/AdviceTest.java:21]: Nit: this value does not seem to be used anywhere

File: api/all/src/main/java/io/opentelemetry/api/metrics/DoubleHistogramBuilder.java [hunk 2]
```diff
@@ -32,6 +34,11 @@ public interface DoubleHistogramBuilder {
    */
   DoubleHistogramBuilder setUnit(String unit);
 
+  /** Specify advice for histogram implementations. */
+  default DoubleHistogramBuilder setAdvice(Consumer<HistogramAdviceConfigurer> adviceConsumer) {
```
Comment (mateuszrzeszutek) [inline api/all/src/main/java/io/opentelemetry/api/metrics/DoubleHistogramBuilder.java:38]: Nit: are we planning to add this to the API right away? Or perhaps introduce it in an `internal` interface first?

**Review by mateuszrzeszutek — APPROVED** [review]
Awesome!

File: api/all/src/main/java/io/opentelemetry/api/metrics/HistogramAdviceConfigurer.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.api.metrics;
+
+import java.util.List;
+
+/** Configure advice for implementations of {@link LongHistogram} and {@link DoubleHistogram}. */
+public interface HistogramAdviceConfigurer {
```
Comment (jkwatson) [inline api/all/src/main/java/io/opentelemetry/api/metrics/HistogramAdviceConfigurer.java:11]: do you envision this having more methods added, or could we add `@FunctionalInterface` to it?

File: api/all/src/main/java/io/opentelemetry/api/metrics/DoubleHistogramBuilder.java (hunk 2 above)
Comment (jack-berg) [reply api/all/src/main/java/io/opentelemetry/api/metrics/DoubleHistogramBuilder.java:38]: I'm open to different options. Could make it an internal interface, or add it to the [api incubator](https://github.com/open-telemetry/opentelemetry-java/tree/main/extensions/incubator), or wait on merging until the spec is marked stable. 

I think I like the idea of an internal interface the best.

Comment (jack-berg) [reply api/all/src/main/java/io/opentelemetry/api/metrics/DoubleHistogramBuilder.java]: I've split out the interfaces to `:extensions:incubator`.

To use, add a dependency on `opentelemetry-extension-incubator`, and cast the histogram builder to the extended version of the interface:
-`((ExtendedDoubleHistogramBuilder) meter.histogramBuilder("histogram")).setAdvice(...)`
-`((ExtendedLongHistogramBuilder) meter.histogramBuilder("histogram").ofLongs()).setAdvice(...)`

**Comment by jack-berg** [comment]
Marked this PR as ready for review. The idea has reasonable support in the specification, and I will continue to drive it. In the meantime, I've adjusted the PR to move the advice APIs to `opentelemetry-extension-incubator`, so its actually in a reasonable state to merge.

**Review by mateuszrzeszutek — APPROVED** [review]
This is awesome :100::rocket:

**Comment by jack-berg** [comment]
Hey @open-telemetry/java-instrumentation-maintainers if the advice API were to be merged in this form, in which its exposed in the extensions API component, would we be able to use it in instrumentation? Also, would we be comfortable using an experimental API in instrumentation?

I think it would be really useful in cases like the gc duration histogram metric, which would benefit a lot from tuning the buckets to match the duration we typically expect for garbage collection events.

**Comment by mateuszrzeszutek** [comment]
> would we be able to use it in instrumentation?

I think so; we'll just have to put `extensions` on the bootstrap classpath. 

> Also, would we be comfortable using an experimental API in instrumentation?

Personally: as long as we're not exposing this in any public API, sure :100:. I've been hoping to use the advice API in the micrometer bridge for a while.

**Review by jkwatson — APPROVED** [review]

**Pull request merged**